v2/tools/generator/internal/codegen/pipeline/convert_allof_and_oneof_to_objects.go

```diff
@@ -581,58 +740,116 @@ func (s synthesizer) handleAllOfType(leftAllOf *astmodel.AllOfType, right astmod
 func (s synthesizer) handleOneOf(leftOneOf *astmodel.OneOfType, right astmodel.Type) (astmodel.Type, error) {
-	// if there is an equal case, use that:
-	{
-		var result astmodel.Type
-		leftOneOf.Types().ForEach(func(lType astmodel.Type, _ int) {
-			if astmodel.TypeEquals(lType, right) {
-				result = lType
-			}
-		})
-
-		if result != nil {
-			return result, nil
-		}
+	// if there is an equal case, use that
+	if leftOneOf.Types().Contains(right, astmodel.EqualityOverrides{}) {
+		return right, nil
 	}
 
 	// otherwise intersect with each type:
-	var newTypes []astmodel.Type
+	newTypes := astmodel.MakeTypeSet()
 	err := leftOneOf.Types().ForEachError(func(lType astmodel.Type, _ int) error {
 		newType, err := s.intersectTypes(lType, right)
 		if err != nil {
 			return err
 		}
 
-		newTypes = append(newTypes, newType)
+		newTypes.Add(newType)
 		return nil
 	})
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "unable to intersect oneOf with %s", astmodel.DebugDescription(right))
 	}
 
-	return astmodel.BuildOneOfType(newTypes...), nil
+	if only, ok := newTypes.Single(); ok {
+		return only, nil
+	}
+
+	// Still have a OneOf, need to reprocess it
+	oneOf := leftOneOf.WithTypes(newTypes.AsSlice())
+	return s.oneOfToObject(oneOf)
 }
 
 func (s synthesizer) handleTypeName(leftName astmodel.TypeName, right astmodel.Type) (astmodel.Type, error) {
-	if found, ok := s.defs[leftName]; !ok {
+	found, ok := s.defs[leftName]
+	if !ok {
 		return nil, errors.Errorf("couldn't find type %s", leftName)
-	} else {
-		result, err := s.intersectTypes(found.Type(), right)
-		if err != nil {
-			return nil, err
-		}
+	}
+
+	if s.activeNames.Contains(leftName) {
+		// Avoid recursion, just keep the existing TypeName
+		return leftName, nil
+	}
 
-		// TODO: can we somehow process these pointed-to types first,
-		// so that their innards are resolved already and we can retain
-		// the names?
+	// Track which TypeNames we're processing, so we can detect recursion
+	s.activeNames.Add(leftName)
+	defer func() { s.activeNames.Remove(leftName) }()
 
-		if astmodel.TypeEquals(result, found.Type()) {
-			// if we got back the same thing we are referencing, preserve the reference
-			return leftName, nil
+	result, err := s.intersectTypes(found.Type(), right)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"couldn't intersect %s with %s", leftName, astmodel.DebugDescription(right))
+	}
+
+	// TODO: can we somehow process these pointed-to types first,
+	// so that their innards are resolved already and we can retain
+	// the names?
+
+	// If no change to the type, keep the typename
+	if astmodel.TypeEquals(result, found.Type()) ||
+		astmodel.TypeEquals(result, leftName) {
+		// if we got back the same thing we already have, preserve the reference
+		return leftName, nil
+	}
+
+	// if this is the only reference to a type we've merged, we can redefine it
+	if refs, ok := s.referenceCounts[leftName]; ok && refs == 1 {
```

Comment:
Also - can you explain here why the type needs to be be redefined? What's the benefit of doing it? What happens when it isn't done (because of other references)?

Reply:
Added comment and inverted condition.